### PR TITLE
Feature/FPV-46 Implement API endpoints for results

### PR DESCRIPTION
**Pull request by kendrick-807**
Implemented two GET routes: '/all' to get the voting result and '/candidateId=:id' to get the voting result for a specific candidate

### CODE REVIEW

File: routes/candidateRoutes.js [hunk 1]
```diff
@@ -47,7 +47,7 @@ candidateRouter.get('/id=:id', authorizedOrdinaryUser, async (req, res) => {
     }
 });
 
-candidateRouter.get('/all', async (req, res) => {
+candidateRouter.get('/all', authorizedOrdinaryUser, async (req, res) => {
```
Comment (phamduylong) [inline routes/candidateRoutes.js]: This was not a mistake, it was intentionally left so. Candidates information are usually public.

File: routes/resultRoutes.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+const express = require('express');
+const resultRouter = express.Router();
+const User = require('../models/User');
+const Candidate = require('../models/Candidate');
+const { authorizedOrdinaryUser } = require('../utilities/utilities');
+const logger = require('../utilities/logger');
+
+resultRouter.get('/all', authorizedOrdinaryUser, async (req, res) => {
```
Comment (phamduylong) [inline routes/resultRoutes.js]: auth is unnecessary

File: routes/resultRoutes.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        logger.info("Voting result successfully obtained");
+        res.status(200).json(response);
+    } catch (error) {
+        logger.error("An error occurred while getting the votes, details:", error);
+        res.status(500).json({ error: error });
+    }
+});
+
+resultRouter.get('/candidateId=:id', authorizedOrdinaryUser, async (req, res) => {
```
Comment (phamduylong) [inline routes/resultRoutes.js]: auth is unnecessary

**Review by phamduylong — CHANGES_REQUESTED** [review]
These are public information

File: routes/candidateRoutes.js (hunk 1 above)
Comment (kendrick-807) [reply routes/candidateRoutes.js]: Okay, I thought it was my mistake. Agreed

File: routes/resultRoutes.js (hunk 2 above)
Comment (kendrick-807) [reply routes/resultRoutes.js]: Agreed.

File: routes/resultRoutes.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    logger.info('Requested result for candidate with ID:', candidateId);
+
+    if (!candidateId || isNaN(candidateId) || candidateId < 0) {
+        logger.error(`Invalid candidate ID: ${candidateId}`);
+        return res.status(400).send({ error: "Invalid candidate ID. Please provide a valid candidate ID." });
+    }
+
+    try {
+        // Get all users who voted for the specified candidate
+        const votedUsers = await User.find({ candidateVotedId: parseInt(candidateId) });
```
Comment (phamduylong) [inline routes/resultRoutes.js]: ~~parseInt(candidateId)~~
```js
const castedId = Number(candidateId);
if(castedId === NaN) // return 400 
const votedUsers = await User.find({ candidateVotedId: castedId });
```

File: routes/resultRoutes.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            const response = {
+                id: candidateInfo.id,
+                name: candidateInfo.name,
+                votesReceived: votedUsers.length,
+            };
+            logger.info(`Result found for candidate. ID: ${candidateInfo.id}, name: ${candidateInfo.name}, votesReceived: ${votedUsers.length}`);
+            return res.status(200).json(response);
+        }
+        logger.info(`Candidate with ID ${candidateId} does not exist!`);
+        res.status(400).json({ error: `Candidate with ID ${candidateId} does not exist!` });
```
Comment (phamduylong) [inline routes/resultRoutes.js]: 404

File: routes/resultRoutes.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+});
+
+resultRouter.get('/candidateId=:id', async (req, res) => {
+    const candidateId = req.params.id;
+    logger.info('Requested result for candidate with ID:', candidateId);
+
+    if (!candidateId || isNaN(candidateId) || candidateId < 0) {
+        logger.error(`Invalid candidate ID: ${candidateId}`);
+        return res.status(400).send({ error: "Invalid candidate ID. Please provide a valid candidate ID." });
```
Comment (phamduylong) [inline routes/resultRoutes.js]: 404

**Review by phamduylong — CHANGES_REQUESTED** [review]

File: routes/resultRoutes.js (hunk 4 above)
Comment (kendrick-807) [reply routes/resultRoutes.js]: 螢幕截圖 2023-11-07 下午2 32 35

Comment (kendrick-807) [reply routes/resultRoutes.js]: and also should this return 404 or 400

Comment (phamduylong) [reply routes/resultRoutes.js]: Difference: 
400: Bad request, i.e sending a string as id, so request was malformed.
404: Not found, i.e id was indeed a number but found no matching candidate with such id

Comment (kendrick-807) [reply routes/resultRoutes.js]:
```js
if (!candidateId || isNaN(candidateId) || candidateId < 0)
```
Isn't candidateId < 0 a kind of bad request, as the id will be always >= 0?

Comment (phamduylong) [reply routes/resultRoutes.js]: Yes, go ahead and check that if you wish to, I think it's fine. **isNaN(candidateId)** will always be false, FYI

Comment (kendrick-807) [reply routes/resultRoutes.js]: Ok, I will leave it then. Actually, if candidateId is a string, **isNaN(candidateId)** will return true

**Review by phamduylong — APPROVED** [review]

**Pull request merged**